File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        charCnt = defaultdict(int)
        L = 0
        N=len(s)
        best = 0
        maxF = 0
        for R in range(N):
            charCnt[s[R]] += 1
            maxF = max(maxF, charCnt[s[R]])
            #windowSize = R - L + 1
            #mostFrequentCnt = max(charCnt.values())
            #Number of chars to change = R - L + 1 - max(charCnt.values()
            while R - L + 1 - maxF > k:
                charCnt[s[L]] -= 1
                L += 1
            best = max(best, R - L +1)
        return best

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int freq [] = new int [26];
        int n = s.length(),l = 0, maxF = 0, res = 0;
        for (int r = 0; r < n; r++) {
            freq[Character.toUpperCase(s.charAt(r)) - 'A']++;
            maxF = Math.max(maxF, freq[Character.toUpperCase(s.charAt(r)) - 'A']);
            if (r - l + 1 - maxF > k) {
                freq[s.charAt(l++)-'A']--;
            }
            res = Math.max(res, r-l+1);
        }
        return res;
    }
}